### PR TITLE
feat: FE - SmartDocuments form functions tests

### PR DESCRIPTION
**Pull request by infonl-marcel**
FE - SmartDocuments form tests

- Added tests for all array functions
- Redone function flattenNestedGroups since old one flattend but as well left nested branches in its result
- Removed redundant flatten function for its newer and leaner brother
- Stripped the flattened groups from group(s) that don't have any templates (array length === 0). In case of editing SmartDocuments form of no influence since ignored anyway, BUT (!!) in case of Document create (from Zaak or Taak) now not showing a Group anymore that has no Templates to show. This for sure will be a future finding as it is now one of my findings!

Solves PZ-4618

### CODE REVIEW

File: src/main/app/src/app/admin/smart-documents.service.ts [hunk 1]
```diff
@@ -9,6 +9,18 @@ import { FoutAfhandelingService } from "../fout-afhandeling/fout-afhandeling.ser
 import { ZacHttpClient } from "../shared/http/zac-http-client";
 import { GeneratedType } from "../shared/utils/generated-types";
 
+export type SmartDocumentsTemplateWithParentId =
```
Comment (bas-info-nl) [inline src/main/app/src/app/admin/smart-documents.service.ts:12]: What do we need this type for? Is it not generated and should therefore not be committed to the code ?

File: src/main/app/src/app/admin/smart-documents.service.ts [hunk 2]
```diff
@@ -192,28 +128,24 @@ export class SmartDocumentsService {
...
-        name: item.name,
-        templates: item.templates,
-      };
+  flattenNestedGroups(
+    groups: MappedSmartDocumentsTemplateGroupWithParentId[],
+  ): MappedSmartDocumentsTemplateFlattenedGroupWithParentId[] {
+    const result = [];
+
+    function traverse(group) {
+      const { id, name, templates = [] } = group;
```
Comment (bas-info-nl) [inline src/main/app/src/app/admin/smart-documents.service.ts:137]: I don't get why we only grab these three properties and not the `groups`. That is used a bit further down directly from the `group`...

**Review by bas-info-nl — APPROVED** [review]

File: src/main/app/src/app/admin/smart-documents.service.ts (hunk 1 above)
Comment (infonl-marcel) [reply src/main/app/src/app/admin/smart-documents.service.ts:12]: It is used for creating type `SmartDocumentsTemplateGroupWithParentId`. But I can also do that inline in that type def:

`export type SmartDocumentsTemplateGroupWithParentId = Omit<
  GeneratedType<"RestSmartDocumentsTemplateGroup">,
  "groups" | "templates"
> & {
  groups?: SmartDocumentsTemplateGroupWithParentId[] | null;
  templates?:
    | (GeneratedType<"RestSmartDocumentsTemplate"> & {
        parentGroupId?: string | null;
      })
    | null;
};`

What is your pref?

Comment (bas-info-nl) [reply src/main/app/src/app/admin/smart-documents.service.ts:12]: what does it do with the `Generated` type? Or is it copying the type definition and leaving information out? I don't like either way then.... I then prefer that we define it normally with the fields that we know we need in this type !

File: src/main/app/src/app/admin/smart-documents.service.ts (hunk 2 above)
Comment (bas-info-nl) [reply src/main/app/src/app/admin/smart-documents.service.ts:137]: so the reason is that these 3 fields are used separately from the `groups` field, and this way that is more obvious? I don't agree with that observation. If we use all these fields, let's get them consistently from the `group` that's passed in, maybe destruct it straight in the function parameter definition? Then we don't even need `group` any more.

Comment (bas-info-nl) [reply src/main/app/src/app/admin/smart-documents.service.ts:137]: Just trying this out locally and I also see there is no type-safety on `result` nor on the `traverse` method.... is that normal? And why?

Comment (bas-info-nl) [reply src/main/app/src/app/admin/smart-documents.service.ts:137]: I would go for this `traverse` method definition:
```
function traverse({ id, name, templates = [], groups }: MappedSmartDocumentsTemplateGroupWithParentId) {
```

Comment (bas-info-nl) [reply src/main/app/src/app/admin/smart-documents.service.ts:137]: I suggest the following:
```suggestion
function traverse({ id, name, templates = [], groups }: MappedSmartDocumentsTemplateGroupWithParentId) {
```

File: src/main/app/src/app/admin/smart-documents.service.ts (hunk 1 above)
Comment (bas-info-nl) [reply src/main/app/src/app/admin/smart-documents.service.ts:12]: I want @xiduzo and @edgarvonk to tell me if we should have this coupling between backend API type and frontend internal type. What is are their opinions?

**Pull request merged**